### PR TITLE
code review !!!

### PR DESCRIPTION
**Pull request by oshinowo-stephen**

### CODE REVIEW

File: config/settings.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
-  "prefix": "##",
+  "prefix": ".",
   "status": {
-    "name": "with your granddads clothes",
+    "name": "just Tagger beta",
```
Comment (alex-taxiera) [inline config/settings.json]: undo these

File: package.json [hunk 2]
```diff
@@ -21,6 +21,8 @@
     "eslint-plugin-import": "^2.14.0",
     "eslint-plugin-node": "^8.0.0",
     "eslint-plugin-promise": "^4.0.1",
-    "eslint-plugin-standard": "^4.0.0"
+    "eslint-plugin-standard": "^4.0.0",
+    "imgur": "^0.3.1",
+    "nodemon": "^1.18.8"
```
Comment (alex-taxiera) [inline package.json]: nodemon does not need to be a dep

File: src/commands/tag.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 const { Command } = require('eris-boiler')
 const { get } = require('superagent')
 
-const IMAGE_REGEXP = new RegExp(/^(https|http):?\/(.*).(png|jpeg|jpg|gif)/)
+const IMAGE_REGEXP = new RegExp(/^(https|http):?\/(.*).(png|jpeg|jpg|gif|PNG|JPEG|JPG|GIF)/)
```
Comment (alex-taxiera) [inline src/commands/tag.js]: why don't we just lowercase out strings before running the regex test?

File: src/commands/tag.js [hunk 4]
```diff
@@ -19,6 +19,7 @@ module.exports = (bot) => new Command(
       const tag = await bot.tag.getTag(key)
       if (!tag) return `Tag \`${key}\` doesn't exists`
 
+      console.log(tag)
```
Comment (alex-taxiera) [inline src/commands/tag.js]: remove this

File: src/commands/tag.js [hunk 5]
```diff
@@ -39,27 +40,18 @@ const add = (bot) => new Command(
     },
     run: async function ({ bot, msg, params }) {
       const [key, url] = params
-      const src = msg.attachments.length > 0 ? msg.attachments[0].url : url
+      const image = msg.attachments.length > 0 ? msg.attachments[0].url : url
```
Comment (alex-taxiera) [inline src/commands/tag.js]: this variable is still a url so src is a better name for it

File: src/tagger/imgurClient.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+const imgur = require('imgur')
+
+const TaggerDatabaseManager = require('./TagDatabaseManager')
+
+class ImgurClient {
+  constructor () {
+    this.tdbm = new TaggerDatabaseManager()
```
Comment (alex-taxiera) [inline src/tagger/imgurClient.js]: just use the instance from the tagger client, we don't need so many database managers

File: src/tagger/imgurClient.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    imgur.setClientId(process.env.IMGUR_CLIENT_ID)
+    imgur.setAPIUrl('https://api.imgur.com/3/')
+  }
+
+  async upload (id, imgObj) {
```
Comment (alex-taxiera) [inline src/tagger/imgurClient.js]: imgObj can be called tag and have a key and a src like everywhere else

File: src/tagger/imgurClient.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    const { data } = await imgur.uploadUrl(imgObj.image)
+
+    try {
```
Comment (alex-taxiera) [inline src/tagger/imgurClient.js]: don't need async await here we can just return the promise and use .then and .catch

File: src/tagger/imgurClient.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      await this.tdbm.addTag(id, imgObj.key, data.link)
+    } catch (err) {
+      return err
+    }
+  }
+
+  async update (key, imgObj) {
```
Comment (alex-taxiera) [inline src/tagger/imgurClient.js]: same as previous func

File: src/tagger/imgurClient.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
+const imgur = require('imgur')
+
+const TaggerDatabaseManager = require('./TagDatabaseManager')
+
+class ImgurClient {
```
Comment (alex-taxiera) [inline src/tagger/imgurClient.js]: maybe call it ImgurManager, client is not a good name

File: src/commands/tag.js [hunk 11]
```diff
@@ -39,27 +40,18 @@ const add = (bot) => new Command(
...
-      }
+      if (!image) return 'Please attach an image or specify a url!'
+      if (!IMAGE_REGEXP.test(image)) return 'Not an image!'
 
-      return bot.tag.addTag(msg.author.id, key, src)
+      return bot.imgur.upload(msg.author.id, { key, image })
         .then(() => `Added tag \`${key}\``)
-        .catch((error) => {
-          switch (error.code) {
+        .catch((err) => {
```
Comment (alex-taxiera) [inline src/commands/tag.js]: leave it as error not err

File: src/commands/tag.js [hunk 12]
```diff
@@ -98,7 +90,7 @@ const update = (bot) => new Command(
     },
     run: async function ({ bot, msg, params }) {
       const [key, url] = params
-      const src = msg.attachments.length > 0 ? msg.attachments[0].url : url
+      let src = msg.attachments.length > 0 ? msg.attachments[0].url : url
```
Comment (alex-taxiera) [inline src/commands/tag.js]: why was this changed from a const?

File: src/tagger/index.js [hunk 13]
```diff
@@ -1,10 +1,23 @@
...
+const ImgurClient = require('./imgurClient')
 
 class TaggerClient extends DataClient {
   constructor (opt = {}) {
     super(opt)
     this.tag = new TagDatabaseManager()
+    this.imgur = new ImgurClient()
+  }
+
+  generateString(length) {
```
Comment (alex-taxiera) [inline src/tagger/index.js]: what is this for?

**Review by alex-taxiera — CHANGES_REQUESTED** [review]
address all comments please

File: src/tagger/imgurClient.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+class ImgurManager {
+  constructor (bot) {
+    this.bot = bot
+
+    imgur.setClientId(process.env.IMGUR_CLIENT_ID)
+    imgur.setAPIUrl('https://api.imgur.com/3/')
+  }
+
+  handleTag (tag) {
```
Comment (alex-taxiera) [inline src/tagger/imgurClient.js]: better name for this function would be `upload` it can still take a Tag

File: src/tagger/imgurClient.js [hunk 15]
```diff
@@ -0,0 +1,16 @@
+const imgur = require('imgur')
+
+class ImgurManager {
+  constructor (bot) {
+    this.bot = bot
+
+    imgur.setClientId(process.env.IMGUR_CLIENT_ID)
```
Comment (alex-taxiera) [inline src/tagger/imgurClient.js]: can we make imgur a property of the manager? like `this._api = imgur`

File: src/commands/tag.js [hunk 16]
```diff
@@ -43,25 +43,18 @@ const add = (bot) => new Command(
...
-              console.error(error)
-              return 'Something went wrong adding this tag, sorry!'
-          }
-        })
+      if (result.data) {
+        try {
+          await bot.tag.addTag(msg.author.id, key, result.data.link)
+          return `Added \`${key}\``
+        } catch (error) {
+          return `Tag \`${key}\` already exists`
```
Comment (alex-taxiera) [inline src/commands/tag.js]: check this before trying to upload to imgur

File: src/commands/tag.js [hunk 17]
```diff
@@ -43,25 +43,18 @@ const add = (bot) => new Command(
...
       if (!src) return 'Please attach an image or specify a url!'
-      if (!IMAGE_REGEXP.test(src)) return 'Not an image!'
+      if (!IMAGE_REGEXP.test(src.toLowerCase())) return 'Not an image!'
 
-      try {
-        await get(src)
-      } catch (err) {
-        return 'Bad link!'
-      }
+      const result = await bot.imgur.handleTag({ key, src })
```
Comment (alex-taxiera) [inline src/commands/tag.js]: needs error handling

File: src/commands/tag.js [hunk 18]
```diff
@@ -113,8 +106,16 @@ const update = (bot) => new Command(
...
-      await bot.tag.updateTag(key, src)
-      return `Updated tag \`${key}\``
+      const result = await bot.imgur.handleTag({ key, src })
+
+      if (result.data) {
+        try {
+          await bot.tag.updateTag(key, result.data.link)
+          return `Updated tag \`${key}\``
+        } catch (error) {
+          console.log(error)
```
Comment (alex-taxiera) [inline src/commands/tag.js]: errors should still return something to tell the user an error occured

File: src/tagger/imgurClient.js [hunk 19]
```diff
@@ -0,0 +1,16 @@
+const imgur = require('imgur')
+
+class ImgurManager {
+  constructor (bot) {
+    this.bot = bot
```
Comment (alex-taxiera) [inline src/tagger/imgurClient.js]: doesn't appear like this class is using bot anymore

**Review by alex-taxiera — DISMISSED** [review]

File: src/commands/tag.js (hunk 16 above)
Comment (oshinowo-stephen) [reply src/commands/tag.js]: wait what?

Comment (alex-taxiera) [reply src/commands/tag.js]: I mean check to see if the tag is in the db before doing anything with imgur

**Review by alex-taxiera — APPROVED** [review]
👍

**Pull request merged**